Added arch arm64 to profiler library

sroeper opened the pull request:
closes 4524

The profiler is not working on the Apple Silicon platform, see Issue 4524.
These changes discard old contents of the library for Mac (ppc, i386) because these can't be any longer updated (see also remarks in original file).

Instead this build script builds a universal library for x86_64 and arm64. My tests with NB19 on Apple Silicon where successful.

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

If you're a committer, please label the PR before pressing "Create pull request" so that the right test jobs can run.

mbien commented:
this PR will now produce https://github.com/apache/netbeans/labels/ci%3Adev-build s downloadable from the summary page of the github workflow. However, I don't think they will contain your changes since the profiler artifact is probably something we have to redeploy via:
https://github.com/apache/netbeans/blob/92a4de30ffe35b769544b8783d09d279b22fd375/profiler/lib.profiler/external/binaries-list#L17-L18

so someone on apple silicon would have to test it manually first.

also: i am pretty sure jdk16 means java 1.6, so we might want to remove that from the PR title to avoid confusion.

sroeper commented:
@mbien I renamed the PR by  removing "for jdk16" as suggested.

- I downloaded the artefact "mac" which is a zip file containing the profiler library "libprofilerinterface.jnilib".
- I checked if it contains x86_64 and arm64 code with lipo -info libprofilerinterface.jnilib and lipo reports that both are included.
- I removed libprofilerinterface.jnilib from my Netbeans 19 installation and dropped the library from the downloaded artefact into that place, an test profiling was successful.

So I think it should be fine.

neilcsmith-net commented:
Thanks @sroeper - let's see if we can get this fixed for NB20.

> so someone on apple silicon would have to test it manually first, and if it works, upload it so that this PR can be updated.

I have access to an Apple M1 machine for builds and testing.

The process is going to be more convoluted than that though.  We need a release process for the binaries to incorporate them into the IDE.  It might be better to extract to a separate repository, similar to https://github.com/apache/netbeans-native-launchers  We can possibly use a GH workflow to build the binaries.  Publishing binaries via Maven doesn't seem viable for this given cross-OS build needed, so possibly looking at dist.a.o instead.

Will start a discussion on dev@ about how we handle this (and terminal support - #6012 / #3467 )

sroeper commented:
@neilcsmith-net I am probably not involved enough to see the big picture. What happens if the build is done? Where does the mac.zip artefact come from? Is it build on a mac or just with a Clang compiler which has support for the arch arm64?

At least the build produced a universal binary with both architectures. So my assumption would be that we don't need the binaries for download anymore because they can be produced on the fly during the build. Is this wrong?

Sorry for asking this dump questions but I have not found any description of the build pipeline.

neilcsmith-net commented:
@sroeper the NetBeans IDE needs to be buildable on any OS, and the convenience binaries that ASF distributes are all based on the main binary zip built in a Linux CI build.  Last time I checked, ASF didn't have macOS CI at all anyway.

So, all native binaries need to be released separately then consumed in the build.  It's one of the added complexities that has led to these not being updated much in the past - eg. see problems updating in #2700 

There is a GitHub CI workflow here already for building the profiler binaries https://github.com/apache/netbeans/blob/master/.github/workflows/native-binary-build-lib.profiler.yml  We should probably extract to make releasing easier.

sroeper commented:
@neilcsmith-net Thank you for the insights!
According to https://github.com/apache/netbeans/blob/master/.github/workflows/native-binary-build-lib.profiler.yml the build of the native profiler lib is done on a mac with the latest OS and probably with the latest Xcode installed. So it does not matter if it is an Intel or Apple Silicon based machine to compile for x86 and arm64, because the compiler is able to cross compile for both architectures. This explains why my change works.

One further bit I don't understand is the role of https://netbeans.osuosl.org/binaries/ , which may makes things more complicated than necessary. As far as I understand one have to download the profiler-external-binaries-ASF.zip from GitHub and manually upload it to https://netbeans.osuosl.org/binaries/ with recreation of checksums etc. Wouldn't it be much easier to upload the artefact to maven.apache.org (with proper versioning of course) and then consume it from the ant build with the ant get task?

And yes I agree that a designated project for all native code could make the build more understandable and easier to maintain. At least this would be true for all artefacts which does not have a dependency to the java code (like open_pty etc). But again there should be an automatic way to upload the artefacts (preferable a maven repository).

neilcsmith-net commented:
@sroeper I've opened a discussion on dev@ https://lists.apache.org/thread/08hsrfgn6klcf9bm8t96h6pv0zrysfjm

OSUOSL was intended for third-party binaries consumed in the build, such as the original Oracle CDDL licensed versions of the profiler binaries.  It should not really have been used for the later profiler binaries.  It's a legacy system we're meant to be moving away from.

We cannot use a fully automated process as we have to go through ASF release procedures.  Looking further I'm inclined to keep in this repo and adapt the GitHub workflow to include a source zip, and to release via dist.apache.org rather than Maven - but let's see what other people say.

neilcsmith-net commented:
@sroeper could you possibly verify that the macOS binary inside `profiler-external-binaries-ASF` at https://github.com/apache/netbeans/actions/runs/6328665359 is working the same for you?  It's from #6496 which incorporates your change and built by the GitHub actions workflow.

sroeper commented:
@neilcsmith-net Sorry for the delay. Verified, the profiler-external-binaries-ASF works on my side.
